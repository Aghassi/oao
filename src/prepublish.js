// @flow

import path from 'path';
import { merge } from 'timm';
import semver from 'semver';
import { mainStory, chalk } from 'storyboard';
import { readAllSpecs, ROOT_PACKAGE } from './utils/readSpecs';
import writeSpecs from './utils/writeSpecs';
import { cp } from './utils/shell';

type Options = {
  src: string,
  ignoreSrc?: string,
  copyAttrs: string,
};

const run = async ({ src, ignoreSrc, copyAttrs: copyAttrsStr }: Options) => {
  const allSpecs = await readAllSpecs(src, ignoreSrc);
  const pkgNames = Object.keys(allSpecs);
  const rootSpecs = allSpecs[ROOT_PACKAGE].specs;

  // Check version numbers!
  const masterVersion = rootSpecs.version;
  for (let i = 0; i < pkgNames.length; i++) {
    const pkgName = pkgNames[i];
    if (pkgName === ROOT_PACKAGE) continue;
    const { specs } = allSpecs[pkgName];
    const { version } = specs;
    if (specs.private) continue;
    if (!semver.valid(version)) {
      mainStory.error(`Invalid version for ${chalk.bold(pkgName)}: ${chalk.bold(version)}`);
      throw new Error('INVALID_VERSION');
    }
    if (semver.gt(version, masterVersion)) {
      mainStory.error(`Version for ${pkgName} (${chalk.bold(version)}) > master version (${chalk.bold(masterVersion)})`);
      throw new Error('INVALID_VERSION');
    }
  }

  // Copy READMEs to all non-private packages
  for (let i = 0; i < pkgNames.length; i++) {
    const pkgName = pkgNames[i];
    if (pkgName === ROOT_PACKAGE) continue;
    const { pkgPath, specs } = allSpecs[pkgName];
    if (specs.private) continue;
    const srcFile = pkgName === rootSpecs.name ? 'README.md' : 'README-LINK.md';
    const dstFile = path.join(pkgPath, 'README.md');
    cp(srcFile, dstFile);
  }

  // Merge common attributes with submodules
  const commonSpecs = {};
  const copyAttrs = copyAttrsStr.split(/\s*,\s*/);
  copyAttrs.forEach((attr) => { commonSpecs[attr] = rootSpecs[attr]; });
  mainStory.info('Updating package attributes', { attach: commonSpecs });
  for (let i = 0; i < pkgNames.length; i++) {
    const pkgName = pkgNames[i];
    if (pkgName === ROOT_PACKAGE) continue;
    const { specPath, specs: prevSpecs } = allSpecs[pkgName];
    if (prevSpecs.private) continue;
    const nextSpecs = merge(prevSpecs, commonSpecs);
    writeSpecs(specPath, nextSpecs);
  }

  mainStory.warn('Please make sure you commit all changes before you attempt "oao publish"');
};

export default run;
