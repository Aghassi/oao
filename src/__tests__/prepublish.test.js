/* eslint-env jest */
/* eslint-disable global-require, import/no-dynamic-require */

import path from 'path';
import prepublish from '../prepublish';

jest.mock('../utils/shell');
jest.mock('../utils/writeSpecs');

const COPY_SPECS = [
  'description', 'keywords',
  'author', 'license',
  'homepage', 'bugs', 'repository',
];

const normalizePath = (p) => p.split(path.sep).join('/');

describe('PREPUBLISH command', () => {
  beforeEach(() => { jest.resetAllMocks(); });

  it('copies READMEs as appropriate', async () => {
    const helpers = require('../utils/shell');
    await prepublish({ src: 'test/fixtures/packages/*' });
    const normalizedArgs = helpers.cp.mock.calls.map(([src, dst]) =>
      [normalizePath(src), normalizePath(dst)]
    );
    expect(normalizedArgs).toMatchSnapshot();
  });

  it('copies common attributes to subpackages', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    const refSpecs = require(path.join(process.cwd(), 'package.json'));
    await prepublish({ src: 'test/fixtures/packages/*' });
    expect(writeSpecs.mock.calls.map((args) => args[1].name)).toEqual(['oao', 'oao-b', 'oao-c', 'oao-d']);
    writeSpecs.mock.calls.forEach(([, specs]) => {
      COPY_SPECS.forEach((attr) => { expect(specs[attr]).toEqual(refSpecs[attr]); });
    });
  });

  it('throws when a package has a version > master', async () => {
    try {
      await prepublish({ src: 'test/fixtures/packagesWrongVersion/*' });
      throw new Error('DID_NOT_THROW');
    } catch (err) {
      if (err.message === 'DID_NOT_THROW') throw err;
    }
  });

  it('throws when a package has an invalid version', async () => {
    try {
      await prepublish({ src: 'test/fixtures/packagesWrongVersion2/*' });
      throw new Error('DID_NOT_THROW');
    } catch (err) {
      if (err.message === 'DID_NOT_THROW') throw err;
    }
  });
});
