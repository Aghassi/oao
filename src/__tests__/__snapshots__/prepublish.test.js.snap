// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PREPUBLISH command copies READMEs as appropriate 1`] = `
Array [
  Array [
    "README.md",
    "test/fixtures/packages/oao/README.md",
  ],
  Array [
    "README-LINK.md",
    "test/fixtures/packages/oao-b/README.md",
  ],
  Array [
    "README-LINK.md",
    "test/fixtures/packages/oao-c/README.md",
  ],
  Array [
    "README-LINK.md",
    "test/fixtures/packages/oao-d/README.md",
  ],
]
`;
